File: 2018/Python/src/01.py
def generator(file) :
    return [int(x) for x in file.splitlines()]

def part_1(input) : 
    return sum(input)

def part_2(input) : 
    visited_frequencies = []
    current = 0
    i = 0
    while current not in visited_frequencies :
        visited_frequencies.append(current)
        current += input[i]
        i = (i + 1) % len(input)
    return current

File: 2018/Rust/src/bin/01.rs
aoc2018::main!();

fn generator(input : &str) -> Vec<i32> {
    input
        .lines()
        .map(|value| value.parse::<i32>().unwrap())
        .collect_vec()
}

fn part_1(input : Vec<i32>) -> i32 {
    input
        .iter()
        .sum()
}

fn part_2(input : Vec<i32>) -> i32 {
    input
        .iter()
        .cycle()
        .fold_while((0, HashSet::new()), |(frequence, mut reached_frequences) , change| 
            if !reached_frequences.insert(frequence + change) {
                FoldWhile::Done((frequence + change, reached_frequences))
            } else {
                FoldWhile::Continue((frequence + change, reached_frequences))
            })
            .into_inner()
            .0
        }